README
NAME
    Net::Riak - Interface to Riak

VERSION
    version 0.1701

SYNOPSIS
        # REST interface
        my $client = Net::Riak->new(
            host => 'http://10.0.0.40:8098',
            ua_timeout => 900,
        );

        # Or PBC interface.
        my $client = Net::Riak->new(
            transport => 'PBC',
            host => '10.0.0.40',
            port => 8080
        );

        my $bucket = $client->bucket('blog');
        my $obj    = $bucket->new_object('new_post', {title => 'foo', content => 'bar'});
        $obj->store;

        $obj = $bucket->get('new_post');
        say "title for ".$obj->key." is ".$obj->data->{title};

        # Indexing and searching (REST interface)
        $client->setup_indexing("bucket_name");
        ...adding documents to riak...
        my $response = $client->search(
            index => 'bucket_name',
            q     => 'field:value'
        );

        # Secondary index setup (REST interface)
        my $obj3 = $bucket->new_object('foo3', {...});
        $obj3->add_index('myindex_bin','myvalue' );
        $obj3->add_index('number_int', 1001);
        $obj3->store;

        # Get all keys for a specific index/value pair
        my @keys = $client->index('mybucket', 'myindex_bin', 'myvalue' );

        # Get all keys for a range of index value pairs
        my @keys = $client->index('mybucket', 'number_int', 500, 1500);

        # Removing a secondary index (REST interface)
        my $new_obj = $bucket->get('foo3');
        $new_obj->remove_index('number_int', 1001);
        $new_obj->store;

DESCRIPTION
  ATTRIBUTES
    host
      REST: The URL of the node

      PBC: The hostname of the node

      default 'http://127.0.0.1:8098'

      Note that providing multiple hosts is now deprecated.

    port
      Port of the PBC interface.

    transport
      Used to select the PB protocol by passing in 'PBC'

    prefix
      Interface prefix (default 'riak')

    mapred_prefix
      MapReduce prefix (default 'mapred')

    r R value setting for this client (default 2)

    w W value setting for this client (default 2)

    dw
      DW value setting for this client (default 2)

    client_id
      client_id for this client

    ua_timeout (REST only)
      timeout for LWP::UserAgent in seconds, defaults to 3.

    disable_return_body (REST only)
      Disable returning of object content in response in a store operation.

      If set to true and the object has siblings these will not be available
      without an additional fetch.

      This will become the default behaviour in 0.17

METHODS
  bucket
        my $bucket = $client->bucket($name);

    Get the bucket by the specified name. Since buckets always exist, this
    will always return a Net::Riak::Bucket

  is_alive
        if (!$client->is_alive) {
            ...
        }

    Check if the Riak server for this client is alive

  all_buckets
    List all buckets, requires Riak 0.14+ or PBC connection.

  add
        my $map_reduce = $client->add('bucket_name', 'key');

    Start assembling a Map/Reduce operation

  link
        my $map_reduce = $client->link();

    Start assembling a Map/Reduce operation

  map
        my $map_reduce = $client->add('bucket_name', 'key')->map("function ...");

    Start assembling a Map/Reduce operation

  reduce
        my $map_reduce = $client->add(..)->map(..)->reduce("function ...");

    Start assembling a Map/Reduce operation

  server_info (PBC only)
        $client->server_info->{server_version};

  stats (REST only)
        say Dumper $client->stats;

  search (REST only)
        $client->search( index => 'bucket_name', q => 'field:value' );

    Makes a query to the index (see Net::Riak::Search for more details on
    parameters)

  setup_indexing (REST only)
        $client->setup_indexing('bucket_name');

    Define precommit hook in order to enable indexing documents written into
    the given bucket

SEE ALSO
    Net::Riak::MapReduce

    Net::Riak::Object

    Net::Riak::Bucket

AUTHOR
    franck cuny <franck@lumberjaph.net>, robin edwards <robin.ge@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by linkfluence.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
0.1701 Fri  1 Mar 2013 13:03:44 GMT
    - Add support for secondary indexes (Theo bot)
    - Remove deprecations for 0.17 release (Robin Edwards)
0.1700 Mon 18 Feb 2013 13:06:57 GMT
    - Add ssl support (Theo bot)
0.1600 Thu Mar 15 15:30:47 GMT 2012
    - catch errors in reading sockets (Damien Krotkine)
    - added Riak search (Germain Maurice)
    - fixed package name (David Hull)
    - various documentation updates

0.1502 Thu Jun 23 15:39:10 BST 2011
    - added status line to REST response (Robin Edwards)

0.1501 Thu May 19 12:07:51 BST 2011
    - fix pod error in Net/Riak.pm
    - fix GH-12 in Map/Reduce operations (franck cuny)

0.15 Wed 20 Apr 2011 14:52:22 BST 
    - fixed link encoding (Simon Wistow)
    - added stats method for REST
    - added server_info method for PBC
    - added list bucket support for both protocols (0.14+ for REST)
    - added PBC support
    - added Test::Riak for running tests on both protocols
    - added disable_return_body flag, in REST mode prevents body being returned on store
    - create new object without keys (SymKat <symkat@symkat.com>) (REST only)

    DEPRECATIONS
    - $object->status please call $object->client->status (for REST only)
    - $object->count_links, added ->has_links and ->all_links
    - disable_return_body will become the default in 0.17
    - support for providing multiple hosts when connecting (only supported by REST)
      all load balancing code to be moved out of Net::Riak by 0.17.

0.14 Mon Mar 14 08:59:10 GMT 2011
    - fixed delete link (Simon Wistow)

0.13 Tue, 08 Feb 2011 16:40:07 GMT
    - Enviroment variable for riak when running tests (Robin Edwards)
    - Connection caching (Andrew Sayers)
    - Automatic key generation support (Simon Wistow)

0.12
    - fix _build_link method (reported by gphat)
    - added useragent timeout parameter (Robin Edwards)
    - added expected status message (Robin Edwards)
    - added delete_object method to Net::Riak::Bucket (Robin Edwards)
    - support for adding bucket as objects in add method (Robin Edwards)

0.11 Wed 03 Nov 2010 03:01:10 PM CET  
    - update POD (franck cuny)
    - fix wrong clearer name (_clear_siblings instead of _clear_links)

0.10 Tue 02 Nov 2010 03:34:16 PM CET  
    - add support of array arguments in Net::Riak::MapReduce->add() (Germain Maurice)
    - add 204 to Object::store expected status codes (Gavin Carr)
    - refactor build link logic (Ben Hengst)
    - requests and responses now stored for debugging (Robin Edwards)
    - documentation updated (Robin Edwards)

0.09 Tue 05 Oct 2010 12:03:59 PM CEST  
    - typo
    - when executing a map/reduce, if the timeout of the map/reduce is higher
      than the value for the useragent's timeout, change for this query the
      value of the useragent timeout to be the same as the one for the job.
    - add key callbacks to Bucket get_properties/get_keys stream mode 
      (Gavin Carr)

0.08 Mon 06 Sep 2010 10:52:15 AM CEST  
    - allow to store non-json object (Gavin Carr)
    - increase LWP's MaxLineLength (Gavin Carr)
    - better error reporting in Buckets (Gavin Carr)

0.07 Tue 10 Aug 2010 10:32:32 AM CEST
    - test shouldn't be run by user

0.06 Tue 10 Aug 2010 10:18:59 AM CEST
    - fix bug in links retriving (thanks to Al Tobey)

0.05 Wed 30 Jun 2010 10:54:52 AM CEST
    - fix vlock

0.04 Wed June 16 2010
    - add get_keys to bucket to get the list of existing keys in a
      bucket
    - add map/reduce examples
    - accept multiple hosts
    - use json::{false,true} in allow_mult

0.03 Mon 14 Jun 2010 05:58:48 PM CEST
    - add roles to generate attributes (mx::r::parameterized)

0.02 Mon 14 Jun 2010 03:48:43 PM CEST
    - links_to_header is now private
    - add some handles
    - client is moved to a class
    - client is now a N::R::Client
    - add tests for bucket, client, object, links

0.01 Mon 14 Jun 2010 01:08:52 PM CEST
    - initial release (port of the python client)


lib/Net/Riak.pm
package Net::Riak;
{
  $Net::Riak::VERSION = '0.1701';
}

# ABSTRACT: Interface to Riak

use Moose;

use Net::Riak::Client;
use Net::Riak::Bucket;
use Net::Riak::Types Client => { -as => 'Client_T' };

with 'Net::Riak::Role::MapReduce';

has client => (
    is       => 'rw',
    isa      => Client_T,
    required => 1,
    handles  => [qw/is_alive all_buckets server_info stats search index setup_indexing/]
);

sub BUILDARGS {
    my ($class, %args) = @_;

    my $transport = $args{transport} || 'REST';
    my $trait = "Net::Riak::Transport::".$transport;

    my $client = Net::Riak::Client->with_traits($trait)->new(%args);
    $args{client} = $client;
    \%args;
}

sub bucket {
    my ($self, $name) = @_;
    my $bucket = Net::Riak::Bucket->new(name => $name, client => $self->client);
    $bucket;
}

1;

__END__

=pod

=head1 NAME

Net::Riak - Interface to Riak

=head1 VERSION

version 0.1701

=head1 SYNOPSIS

    # REST interface
    my $client = Net::Riak->new(
        host => 'http://10.0.0.40:8098',
        ua_timeout => 900,
    );

    # Or PBC interface.
    my $client = Net::Riak->new(
        transport => 'PBC',
        host => '10.0.0.40',
        port => 8080
    );

    my $bucket = $client->bucket('blog');
    my $obj    = $bucket->new_object('new_post', {title => 'foo', content => 'bar'});
    $obj->store;

    $obj = $bucket->get('new_post');
    say "title for ".$obj->key." is ".$obj->data->{title};

    # Indexing and searching (REST interface)
    $client->setup_indexing("bucket_name");
    ...adding documents to riak...
    my $response = $client->search(
        index => 'bucket_name',
        q     => 'field:value'
    );

    # Secondary index setup (REST interface)
    my $obj3 = $bucket->new_object('foo3', {...});
    $obj3->add_index('myindex_bin','myvalue' );
    $obj3->add_index('number_int', 1001);
    $obj3->store;

    # Get all keys for a specific index/value pair
    my @keys = $client->index('mybucket', 'myindex_bin', 'myvalue' );

    # Get all keys for a range of index value pairs
    my @keys = $client->index('mybucket', 'number_int', 500, 1500);

    # Removing a secondary index (REST interface)
    my $new_obj = $bucket->get('foo3');
    $new_obj->remove_index('number_int', 1001);
    $new_obj->store;

=head1 DESCRIPTION

=head2 ATTRIBUTES

=over 2

=item B<host>

REST: The URL of the node

PBC: The hostname of the node

default 'http://127.0.0.1:8098'

Note that providing multiple hosts is now deprecated.

=item B<port>

Port of the PBC interface.

=item B<transport>

Used to select the PB protocol by passing in 'PBC'

=item B<prefix>

Interface prefix (default 'riak')

=item B<mapred_prefix>

MapReduce prefix (default 'mapred')

=item B<r>

R value setting for this client (default 2)

=item B<w>

W value setting for this client (default 2)

=item B<dw>

DW value setting for this client (default 2)

=item B<client_id>

client_id for this client

=item B<ua_timeout (REST only)>

timeout for L<LWP::UserAgent> in seconds, defaults to 3.

=item B<disable_return_body (REST only)>

Disable returning of object content in response in a store operation.

If set to true and the object has siblings these will not be available without an additional fetch.

This will become the default behaviour in 0.17

=back

=head1 METHODS

=head2 bucket

    my $bucket = $client->bucket($name);

Get the bucket by the specified name. Since buckets always exist, this will always return a L<Net::Riak::Bucket>

=head2 is_alive

    if (!$client->is_alive) {
        ...
    }

Check if the Riak server for this client is alive

=head2 all_buckets

List all buckets, requires Riak 0.14+ or PBC connection.

=head2 add

    my $map_reduce = $client->add('bucket_name', 'key');

Start assembling a Map/Reduce operation

=head2 link

    my $map_reduce = $client->link();

Start assembling a Map/Reduce operation

=head2 map

    my $map_reduce = $client->add('bucket_name', 'key')->map("function ...");

Start assembling a Map/Reduce operation

=head2 reduce

    my $map_reduce = $client->add(..)->map(..)->reduce("function ...");

Start assembling a Map/Reduce operation

=head2 server_info (PBC only)

    $client->server_info->{server_version};

=head2 stats (REST only)

    say Dumper $client->stats;

=head2 search (REST only)

    $client->search( index => 'bucket_name', q => 'field:value' );

Makes a query to the index (see L<Net::Riak::Search> for more details on parameters)

=head2 setup_indexing (REST only)

    $client->setup_indexing('bucket_name');

Define precommit hook in order to enable indexing documents written into the given bucket

=head1 SEE ALSO

L<Net::Riak::MapReduce>

L<Net::Riak::Object>

L<Net::Riak::Bucket>

=head1 AUTHOR

franck cuny <franck@lumberjaph.net>, robin edwards <robin.ge@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by linkfluence.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
